README.md
---
page_type: sample
languages:
- powershell
products:
- azure
- azure-netapp-files
description: "This project demonstrates how to use Powershell with NetApp Files SDK for Microsoft.NetApp resource provider to deploy SMB Volume."
---

# Azure NetAppFiles SDK SMB Sample Powershell

This project demonstrates how to use Powershell with NetApp Files SDK for Microsoft.NetApp resource provider to deploy SMB Volume.

In this sample application we perform the following operations:

* Creation
  * ANF Account
  * Capacity pool 
  * SMB Volume  
 
* Deletion, the clean up process takes place (not enabled by default, please set the parameter CleanupResources to $true if you want the clean up code to take a place),deleting all resources in the reverse order following the hierarchy otherwise we can't remove resources that have nested resources still live.


If you don't already have a Microsoft Azure subscription, you can get a FREE trial account [here](http://go.microsoft.com/fwlink/?LinkId=330212).

## Prerequisites

1. Azure Subscription
1. Subscription needs to be enabled for Azure NetApp Files. For more information, please refer to [this](https://docs.microsoft.com/azure/azure-netapp-files/azure-netapp-files-register#waitlist) document.
1. Resource Group created
1. Virtual Network with a delegated subnet to Microsoft.Netapp/volumes resource. For more information, please refer to [Guidelines for Azure NetApp Files network planning](https://docs.microsoft.com/en-us/azure/azure-netapp-files/azure-netapp-files-network-topologies)
1. Azure PowerShell, please refer to [Install Azure PowerShell](https://docs.microsoft.com/en-us/powershell/azure/install-az-ps?view=azps-4.8.0)
1. Run the following PS command:
	* Install-Module Az.NetAppFiles
1. Active Directory infrastructure setup with one or more DNS servers from the AD domain (usually the Domain Controllers) available in the same virtual network where you're setting up Azure NetApp Files. If you want to setup an Active Directory test environment, please refer to [Create a new Windows VM and create a new AD Forest](https://github.com/Azure/azure-quickstart-templates/tree/master/active-directory-new-domain#create-a-new-windows-vm-and-create-a-new-ad-forest-domain-and-dc), Domain and DC for a quick setup, then you can work on the vnet that gets created to setup the subnet requirements for ANF.

# How the project is structured

The following table describes all files within this solution:

| Folder     | FileName                | Description                                                                                                                         												 |
|------------|-------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| src        | CreateANFVolume.ps      | Authenticates and executes all operations                                                                                           												 |
| src\Common | CommonOperations.psm1   | PowerShell module that exposes some functions to perform creation and deletion Azure NetApp Files resource that calls the actual Az.NetAppFiles cmdlets							 |
| src\Common | Utils.psm1              | Static class that exposes a few methods that helps on various tasks, like writting a log to the console for example.                												 |
| src\Common | AzureAuth.psm1	       | PoweShell module that exposes functions to connect to Azure and choose target subscription                                          												 |

# How to run the PowerShell script

1. Clone it locally
    ```powershell
    git clone https://github.com/Azure-Samples/netappfiles-powershell-nfs-sdk-sample.git
    ```
	
1. Modes: There are two options to run the script

	Basic mode - Single script create Azure NetApp files with no validation.
	 * Change folder to **netappfiles-powershell-nfs-sdk-sample\src\Basic**
	 * Open CreateANFVolume and edit all the parameters
	 * Save and close
	 * Run the following command
	 ``` powershell
	 CreateANFVolume.ps1
	 ```

	OR
	
    Advanced mode - More advanced way to run the script to create Azure Netapp Files with validation using modules
	 * Change folder to **netappfiles-powershell-nfs-sdk-sample\src\Advanced**
	 * Change values bewtween brackets [ ] below and then run the command 
     ```powershell 
	 CreateANFVolume.ps1 -SubscriptionId '[subscriptionId]' -ResourceGroupName '[Azure Resource Group Name]' -Location '[Azure Location]' -NetAppAccountName '[ANF Account Name]' -NetAppPoolName '[ANF Capacity Pool Name]' -ServiceLevel [Ultra,Premium, Standard] -NetAppVolumeName '[ANF Volume Name]' -SubnetId '[Subnet ID] -DomainJoinUsername '[Username]' -DomainJoinPassword '[Password]' -DNSList 'DNS List' -ADFQDN 'Domain Name' -SmbServerNamePrefix 'SMB Server Name'
     ```
	
	>Note: The below table shows all the mandatory and optional parameters
	
	| Parameter  			| Mandatory | Default Value |
	|-----------------------|-----------|---------------|
	| -SubscriptionId   	| Yes		| 				|
	| -ResourceGroupName	| Yes       | 				|
	| -Location 			| Yes       | 				|
	| -NetAppAccountName	| Yes		|				|
	| -NetAppPoolName		| Yes		|				|
	| -ServiceLevel			| Yes		|				|
	| -NetAppPoolSize		| No		| 4398046511104 |
    | -NetAppVolumeName		| Yes		|				|
    | -ProtocolType			| Yes		| 				|
    | -NetAppVolumeSize		| No		| 107374182400	|
    | -SubnetId				| Yes		|				|
    | -DomainJoinUsername	| Yes		|	 			| 
    | -DomainJoinPassword	| Yes		| 				|
    | -DNSList				| Yes		| 				|
	| -ADFQDN             	| Yes		|				|
	| -SmbServerNamePrefix 	| Yes		|				|
    | -CleanupResources		| No		| False			|
	
	
	Sample output
	![e2e execution](./media/e2e-execution.PNG)

# Troubleshoot

If you encounter the below issue when running the PoweShell command

```
.\CreateANFVolume.ps1 : .\CreateANFVolume.ps1 cannot be loaded. The file .\CreateANFVolume.ps1 is not digitally signed. You cannot 
run this script on the current system.
```

Run the following command
Set-ExecutionPolicy -Scope CurrentUser -ExecutionPolicy Bypass

# References

* [Sign in with Azure PowerShell](https://docs.microsoft.com/en-us/powershell/azure/authenticate-azureps?view=azps-4.8.0)
* [Azure PowerShell AZ Module](https://docs.microsoft.com/en-us/powershell/azure/new-azureps-module-az?view=azps-4.8.0)
* [AZ.NetAppFile](https://docs.microsoft.com/en-us/powershell/module/az.netappfiles/?view=azps-4.8.0#netapp-files)
* [Resource limits for Azure NetApp Files](https://docs.microsoft.com/en-us/azure/azure-netapp-files/azure-netapp-files-resource-limits)
* [Azure Cloud Shell](https://docs.microsoft.com/en-us/azure/cloud-shell/quickstart)
* [Download Azure SDKs](https://azure.microsoft.com/downloads/)


CHANGELOG.md
## Azure NetApp Files - SMB Sample PowerShell script

[1.0.0 (2020-10-15)](#1.0.0 (2020-10-15))

# 1.0.0 (2019-10-09)

*Features*
* Initial commit

*Bug Fixes*
* N/A

*Breaking Changes*
* N/A


src/Basic/CreateANFVolume.ps1
# Copyright(c) Microsoft and contributors. All rights reserved
#
# This source code is licensed under the MIT license found in the LICENSE file in the root directory of the source tree

<#
.SYNOPSIS
    This script creates Azure Netapp files resources with SMB volume type
.DESCRIPTION
    Authenticates with Azure and select the targeted subscription first, then created ANF account, capacity pool and SMB Volume
.PARAMETER SubscriptionId
    Target Subscription
.PARAMETER ResourceGroupName
    Name of the Azure Resource Group where the ANF will be created
.PARAMETER Location
    Azure Location (e.g 'WestUS', 'EastUS')
.PARAMETER NetAppAccountName
    Name of the Azure NetApp Files Account
.PARAMETER NetAppPoolName
    Name of the Azure NetApp Files Capacity Pool
.PARAMETER ServiceLevel
    Service Level - Ultra, Premium or Standard
.PARAMETER NetAppPoolSize
    Size of the Azure NetApp Files Capacity Pool in Bytes. Range between 4398046511104 and 549755813888000
.PARAMETER NetAppVolumeName\
    Name of the Azure NetApp Files Volume
.PARAMETER NetAppVolumeSize
    Size of the Azure NetApp Files volume in Bytes. Range between 107374182400 and 109951162777600
.PARAMETER SubnetId
    The Delegated subnet Id within the VNET
.PARAMETER DomainJoinUsername
    Domain Username
.PARAMETER DomainJoinPassword
    Domain Password
.PARAMETER DNSList
    Comma-seperated DNS list
.PARAMETER ADFQDN
    Active Directory FQDN
.PARAMETER SmbServerNamePrefix
    SMB Server name prefix
.PARAMETER CleanupResources
    If the script should clean up the resources, $false by default
.EXAMPLE
    PS C:\\> CreateANFVolume.ps1 -SubscriptionId '00000000-0000-0000-0000-000000000000' -ResourceGroupName 'My-RG' -Location 'WestUS' -NetAppAccountName 'testaccount' -NetAppPoolName 'pool1' -ServiceLevel Standard -NetAppVolumeName 'vol1' -ProtocolType NFSv4.1 -SubnetId '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/My-RG/providers/Microsoft.Network/virtualNetworks/vnet1/subnets/subnet1'
#>
param
(
    # Name of the Azure Resource Group
    [string]$ResourceGroupName = 'My-RG',

    #Azure location 
    [string]$Location = 'WestUS',

    #Azure NetApp Files account name
    [string]$NetAppAccountName = 'anfaccount',

    #Azure NetApp Files capacity pool name    
    [string]$NetAppPoolName = 'pool1',

    # Service Level can be {Ultra, Premium or Standard}
    [ValidateSet("Ultra","Premium","Standard")]
    [string]$ServiceLevel = 'Standard',

    #Azure NetApp Files capacity pool size
    [ValidateRange(4398046511104,549755813888000)]
    [long]$NetAppPoolSize = 4398046511104,

    #Azure NetApp Files volume name
    [string]$NetAppVolumeName = 'vol1',

    #Azure NetApp Files volume size
    [ValidateRange(107374182400,109951162777600)]
    [long]$NetAppVolumeSize=107374182400,

    #Subnet Id 
    [string]$SubnetId = 'Subnet ID',
    
    #Domain Join Username    
    [string]$DomainJoinUsername = 'pmcadmin',

    #Domain Join Password
    [string]$DomainJoinPassword = 'Pa$$w0rd',

    #DNS List
    [string]$DNSList = '10.0.2.4,10.0.2.5',

    #Active Directory FQDN
    [string]$ADFQDN = 'testdomain.local',

    #SMB Server Name Prefix
    [string]$SmbServerNamePrefix = 'pmcsmb',

    #Clean Up resources
    [bool]$CleanupResources = $false
)

[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

# Authorizing and connecting to Azure
Write-Verbose -Message "Authorizing with Azure Account..." -Verbose
Add-AzAccount

# Create Azure NetApp Files Account
Write-Verbose -Message "Creating Azure NetApp Files Account -> $NetAppAccountName" -Verbose
$ActiveDirectory = New-Object Microsoft.Azure.Commands.NetAppFiles.Models.PSNetAppFilesActiveDirectory
$ActiveDirectory.Dns = $DNSList
$ActiveDirectory.Username = $DomainJoinUsername
$ActiveDirectory.Password = $DomainJoinPassword
$ActiveDirectory.Domain = $ADFQDN
$ActiveDirectory.SmbServerName = $SmbServerNamePrefix

$NewANFAccount = New-AzNetAppFilesAccount -ResourceGroupName $ResourceGroupName `
    -Location $Location `
    -Name $NetAppAccountName `
    -ActiveDirectory @($ActiveDirectory)

Write-Verbose -Message "Azure NetApp Files Account has been created successfully: $($NewANFAccount.Id)" -Verbose

# Create Azure NetApp Files Capacity Pool                                                                                                       
Write-Verbose -Message "Creating Azure NetApp Files Capacity Pool -> $NetAppPoolName" -Verbose                                         
$NewANFPool= New-AzNetAppFilesPool -ResourceGroupName $ResourceGroupName `
    -Location $Location `
    -AccountName $NetAppAccountName `
    -Name $NetAppPoolName `
    -PoolSize $NetAppPoolSize `
    -ServiceLevel $ServiceLevel

Write-Verbose -Message "Azure NetApp Files Capacity Pool has been created successfuuly: $($NewANFPool.Id)" -Verbose

#Create Azure NetApp Files NFS Volume
Write-Verbose -Message "Creating Azure NetApp Files - SMB Volume -> $NetAppVolumeName" -Verbose
$NewANFVolume = New-AzNetAppFilesVolume -ResourceGroupName $ResourceGroupName `
    -Location $Location `
    -AccountName $NetAppAccountName `
    -PoolName $NetAppPoolName `
    -Name $NetAppVolumeName `
    -UsageThreshold $NetAppVolumeSize `
    -SubnetId $SubnetId `
    -CreationToken $NetAppVolumeName `
    -ServiceLevel $ServiceLevel `
    -ProtocolType CIFS    

Write-Verbose -Message "Azure NetApp Files Volume has been created successfully: $($NewANFVolume.Id)" -Verbose

Write-Verbose -Message "====> SMB Server FQDN: $($NewANFVolume.MountTargets[0].smbServerFQDN.ToString())" -Verbose

Write-Verbose -Message "Azure NetApp Files has been created successfully." -Verbose

if($CleanupResources)
{
    
    Write-Verbose -Message "Cleaning up Azure NetApp Files resources..." -Verbose

    Write-Verbose -Message "Deleting Azure NetApp Files Volume $NetAppVolumeName" -Verbose
    Remove-AzNetAppFilesVolume -ResourceGroupName $ResourceGroupName `
        -AccountName $NetAppAccountName `
        -PoolName $NetAppPoolName `
        -Name $NetAppVolumeName
    
    Write-Verbose -Message "Deleting Azure NetApp Files Volume $NetAppPoolName" -Verbose
    Remove-AzNetAppFilesPool -ResourceGroupName $ResourceGroupName -AccountName $NetAppAccountName -PoolName $NetAppPoolName
   
    Write-Verbose -Message "Deleting Azure NetApp Files Volume $NetAppAccountName" -Verbose
    Remove-AzNetAppFilesAccount -ResourceGroupName $ResourceGroupName -Name $NetAppAccountName
   
    Write-Verbose -Message "All Azure NetApp Files resources have been deleted successfully." -Verbose       
}
